packages/abi-coder/src/types/JsonAbiNew.ts

```diff
@@ -0,0 +1,101 @@
+/**
+ * Types for Fuel JSON ABI Format as defined on:
+ * https://github.com/FuelLabs/fuel-specs/blob/master/src/abi/json-abi-format.md
+ */
+export interface JsonAbi {
+  readonly specVersion: string;
+  readonly encodingVersion: string;
+  readonly programType: string;
+  readonly concreteTypes: readonly ConcreteType[];
+  readonly metadataTypes: readonly MetadataType[];
+  readonly functions: readonly AbiFunction[];
+  readonly loggedTypes: readonly LoggedType[];
+  readonly messagesTypes: readonly MessageType[];
+  readonly configurables: readonly Configurable[];
+}
+
+export interface ConcreteType {
+  readonly type: string;
+  readonly concreteTypeId: string;
+  readonly metadataTypeId?: number;
+  readonly typeArguments?: readonly string[];
+}
+
+export interface MetadataType {
+  readonly type: string;
+  readonly metadataTypeId: number;
+  readonly components?: readonly Component[];
+  readonly typeParameters?: readonly number[];
```

Comment:
Is this an array of concrete type indexes or metadata type indexes? It's hard to tell which array this refers to.

Reply:
It is an array of metadata type ids:
> "typeParameters": an array of type metadata declaration ID of the type parameters of the type, if the type is generic, otherwise nonexistent. Each type parameter is a type declaration and is represented as described in [Generic Type Parameter](https://github.com/FuelLabs/fuel-specs/blob/master/src/abi/json-abi-format.md#generic-type-parameter).

Reply:
IDs as in the hashes or the array indexes? Everywhere else, it looks like the hashes are stored as strings. Are these hashes stored as numbers instead or are these array indexes?

The spec itself is unclear on this.

Reply:
Metadata type ids are _always_ numbers. On the topic of array indexes, I wouldn't use array indexes for anything; they might correlate with the `metadataTypeId` of an entry in the `metadataTypes` array, but the spec doesn't define that relationship.